### PR TITLE
Document result handling coding standards

### PR DESCRIPTION
**Pull request by kuhar**
Add a new section on result handling to the coding standards doc with
proposal 1. from `[RFC] Improving LLPC Result checking`.

The utilities mentioned in the new section were added in
https://github.com/GPUOpen-Drivers/llpc/pull/1500.

Issue: https://github.com/GPUOpen-Drivers/llpc/issues/1474.

### CODE REVIEW

File: docs/CodingStandards.md [hunk 1]
```diff
@@ -112,6 +112,64 @@ it is a single line, including comments. Otherwise the body should be braced. Fo
...
+return result;
+```
+
+In situations when it is not possible to meaningfully handle an error condition, you can
+wrap the operation with `llpc::mustSucceed`, e.g.:
+```c++
+mustSucceed(openFile(path), Twine("Failed to open: ") + path);
+```
+
+This will print an error message an abort execution (in debug builds) if `openFile(path)`
```
Comment (vettoreldaniele) [inline docs/CodingStandards.md:167]: nit: and

File: docs/CodingStandards.md [hunk 2]
```diff
@@ -112,6 +112,64 @@ it is a single line, including comments. Otherwise the body should be braced. Fo
...
+returns a non-success value. The error message parameter is optional.
+
+In order to ensure that all `Result` return values are checked and handle, you should
```
Comment (vettoreldaniele) [inline docs/CodingStandards.md:170]: nit: handled

File: docs/CodingStandards.md [hunk 3]
```diff
@@ -112,6 +112,64 @@ it is a single line, including comments. Otherwise the body should be braced. Fo
...
+annotate all function declarations with `LLPC_NODISCARD`.
```
Comment (vettoreldaniele) [inline docs/CodingStandards.md:171]: Just to make it clearer, I'd specify "all functions that do return a Result".

**Comment by amdvlk-admin** [comment]
### Test summary for commit 597d67a345a5086c62a54facd31abb6a78551e75
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 6cc0a035259663610287d8d99262505e78e01db2</li>
<li>SPVGEN: dev efa3cec20c6ebc66254ff9cca9b3ea124811d7a3</li>
<li>XGL: dev 089c9244e6baa5ed42b1342b391c7c6aec521a2f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ad9f69a6cf94a03817d1d40bacf1920f31e3c727</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
V1: Apply suggestions from @vettoreldaniele and fix typos.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 96c8c675b5c917341d8971aa0e3753a15e9226ea
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 6cc0a035259663610287d8d99262505e78e01db2</li>
<li>SPVGEN: dev efa3cec20c6ebc66254ff9cca9b3ea124811d7a3</li>
<li>XGL: dev 089c9244e6baa5ed42b1342b391c7c6aec521a2f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ad9f69a6cf94a03817d1d40bacf1920f31e3c727</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

File: docs/CodingStandards.md [hunk 4]
```diff
@@ -112,6 +112,64 @@ it is a single line, including comments. Otherwise the body should be braced. Fo
...
+wrap the operation with `llpc::mustSucceed`, e.g.:
+```c++
+mustSucceed(openFile(path), Twine("Failed to open: ") + path);
+```
+
+This will print an error message and abort execution (in debug builds) if `openFile(path)`
+returns a non-success value. The error message parameter is optional.
+
+In order to ensure that all `Result` return values are checked and handled, you should
+annotate all function declarations that do return a `Result` with `LLPC_NODISCARD`.
```
Comment (Flakebi) [inline docs/CodingStandards.md:171]: Can you add an example for this?

**Review by Flakebi — COMMENTED** [review]
Looks good, I like that style of error handling and I hope that this gives us more meaningful error messages in the future.

**Comment by kuhar** [comment]
V2: Add examples for `LLPC_NODISCARD`.

**Review by Flakebi — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit b8a08eeff7bc6896e6493449b683a6b8293d7f48
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 6cc0a035259663610287d8d99262505e78e01db2</li>
<li>SPVGEN: dev efa3cec20c6ebc66254ff9cca9b3ea124811d7a3</li>
<li>XGL: dev 089c9244e6baa5ed42b1342b391c7c6aec521a2f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ad9f69a6cf94a03817d1d40bacf1920f31e3c727</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Pull request merged**